ffi/orcjit.cpp

```diff
@@ -0,0 +1,337 @@
+#include "core.h"
+#include "llvm-c/LLJIT.h"
+#include "llvm-c/Orc.h"
+#include <sstream>
+
+#include "llvm/AsmParser/Parser.h"
+#include "llvm/ExecutionEngine/Orc/Core.h"
+#include "llvm/ExecutionEngine/Orc/ExecutionUtils.h"
+#include "llvm/ExecutionEngine/Orc/LLJIT.h"
+#include "llvm/ExecutionEngine/Orc/RTDyldObjectLinkingLayer.h"
+#include "llvm/ExecutionEngine/SectionMemoryManager.h"
+#include "llvm/IRReader/IRReader.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/SourceMgr.h"
+using namespace llvm;
+using namespace llvm::orc;
+
+inline LLJIT *unwrap(LLVMOrcLLJITRef P) { return reinterpret_cast<LLJIT *>(P); }
+
+inline TargetMachine *unwrap(LLVMTargetMachineRef TM) {
+    return reinterpret_cast<TargetMachine *>(TM);
+}
+
+inline LLVMOrcJITTargetMachineBuilderRef wrap(JITTargetMachineBuilder *JTMB) {
+    return reinterpret_cast<LLVMOrcJITTargetMachineBuilderRef>(JTMB);
+}
+
+class JITDylibTracker {
+  public:
+    std::shared_ptr<LLJIT> lljit;
+    IntrusiveRefCntPtr<llvm::orc::ResourceTracker> tracker;
+    JITDylib &dylib;
+    JITDylibTracker(std::shared_ptr<LLJIT> &lljit_, JITDylib &dylib_,
+                    IntrusiveRefCntPtr<llvm::orc::ResourceTracker> &&tracker_)
+        : lljit(lljit_), dylib(dylib_), tracker(tracker_) {}
+};
+
+typedef struct {
+    uint8_t element_kind;
+    char *value;
+    size_t value_len;
+} LinkElement;
+typedef struct {
+    char *name;
+    uint64_t address;
+} SymbolAddress;
+extern "C" {
+
+API_EXPORT(std::shared_ptr<LLJIT> *)
+LLVMPY_CreateLLJITCompiler(LLVMTargetMachineRef tm, const char **OutError) {
+    LLJITBuilder builder;
+
+    if (tm) {
+        // The following is based on
+        // LLVMOrcJITTargetMachineBuilderCreateFromTargetMachine. However,
+        // we can't use that directly because it destroys the target
+        // machine, but we need to keep it alive because it is referenced by
+        // / shared with other objects on the Python side.
+        auto *template_tm = unwrap(tm);
+
+        builder.setJITTargetMachineBuilder(
+            JITTargetMachineBuilder(template_tm->getTargetTriple())
+                .setCPU(template_tm->getTargetCPU().str())
+                .setRelocationModel(template_tm->getRelocationModel())
+                .setCodeModel(template_tm->getCodeModel())
+                .setCodeGenOptLevel(template_tm->getOptLevel())
+                .setFeatures(template_tm->getTargetFeatureString())
+                .setOptions(template_tm->Options));
```

Comment:
Numba might need to override these settings. 
- relocmodel and codemodel might need to be change because the default might not always work with JIT.
- optlevel depends on user selected opt-level which can be changed for debugging.
- target features are changed to disable certain CPU features; useful for making generic reusable code cache.

llvmlite's TargetMachine class does not expose these options yet. We might want to add the setters there or just let user pass these settings directly as argument in the current function.

Reply:
Looking at `LLVMPY_CreateTargetMachine` already allows setting most of these and that's bound through `Target.create_target_machine`. It doesn't allow setting all `TargetOptions`, but I think we should extend that as required. I don't see the need for a separate API here.

Reply:
ok